openedx/features/course_experience/views/course_message.py

```diff
@@ -0,0 +1,130 @@
+from django.utils.translation import get_language, to_locale
+from django.utils.translation import ugettext as _
+from openedx.core.djangolib.markup import Text, HTML
+from opaque_keys.edx.keys import CourseKey
+from web_fragments.fragment import Fragment
+
+from openedx.core.djangoapps.plugin_api.views import EdxFragmentView
+from openedx.features.course_experience import CourseHomeMessages
+
+
+class CourseMessageFragmentView(EdxFragmentView):
+    """
+    A fragment that displays a course message with an alert and call
+    to action for three types of users:
+
+    1) Not logged in users are given a link to sign in or register.
+    2) Unenrolled users are given a link to enroll.
+    3) Enrolled users who get to the page before the course start date
+    are given the option to add the start date to their calendar.
+
+    This fragment requires a user_access map as follows:
+
+    user_access = {
+        'is_anonymous': True if the user is logged in, False otherwise
+        'is_enrolled': True if the user is enrolled in the course, False otherwise
+        'is_staff': True if the user is a staff member of the course, False otherwise
+    }
+    """
+    def render_to_fragment(self, request, course_id, user_access, **kwargs):
+        """
+        Renders a course message fragment for the specified course.
+        """
+        course_key = CourseKey.from_string(course_id)
+        course = get_course_with_access(request.user, 'load', course_key)
+
+        # Get time until the start date, if already started, or no start date, value will be zero or negative
+        now = datetime.now(UTC())
+        already_started = course.start and now > course.start
+        days_until_start_string = "started" if already_started else format_timedelta(course.start - now, locale=to_locale(get_language()))
+        course_start_data = {
+            'course_start_date': format_date(course.start, locale=to_locale(get_language())),
+            'already_started': already_started,
+            'days_until_start_string': days_until_start_string
+        }
+
+        # Register the course home messages to be loaded on the page
+        self.register_course_home_messages(request, course, user_access, course_start_data)
+
+        # Grab the relevant messages
+        course_home_messages = list(CourseHomeMessages.user_messages(request))
+
+        # Return None if user is enrolled and course has begun
+        if user_access['is_enrolled'] and already_started:
+            return None
+
+        # Grab the logo
+        image_src = "course_experience/images/xsy_circle.png"
+
+        context = {
+            'course_home_messages': course_home_messages,
+            'image_src': image_src,
+        }
+
+        html = render_to_string('course_experience/course-message-fragment.html', context)
+        return Fragment(html)
+
+    @staticmethod
+    def register_course_home_messages(request, course, user_access, course_start_data):
```

Comment:
Nit: I think I'd prefer this logic to be in ``course_home.py``. The way this is set up now, the messages are not part of the UI, but are more business logic. It seems a little hidden inside this fragment, and in fact this fragment will show any ``CourseHomeMessages`` that are registered in the future. I would prefer that this fragment be very generic and have no knowledge of the particular messages we want to add at the moment.